### PR TITLE
bpo-35155: clarify protocol handler method naming

### PR DESCRIPTION
**Pull request by Denton-L**
Clarify that the naming of protocol handler methods shouldn't be literally called "protocol" but should be named after the actual protocol.

https://bugs.python.org/issue35155

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Our records indicate we have not received your CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

File: Doc/library/urllib.request.rst
Comment (lysnikolaou) [inline Doc/library/urllib.request.rst]: I think that we need two spaces after a period.
```suggestion
   special case).  Note that, in the following, *protocol* should be replaced
```

File: Doc/library/urllib.request.rst
Comment (vadmium) [inline Doc/library/urllib.request.rst]: I think the originals make more sense; _protocol_ is used as a word in these cases.

File: Doc/library/urllib.request.rst
Comment (vadmium) [inline Doc/library/urllib.request.rst:623]: The contrast with _http_error_type_ now makes it less obvious that _type_ is also a placeholder. I suggest to keep the notation uniform.

File: Doc/library/urllib.request.rst
Comment (Denton-L) [reply Doc/library/urllib.request.rst]: I interpreted it to mean that `protocol` was the actual "parameter" which was passed in, rather than protocol in the sense of a word.

**Comment by Denton-L** [comment]
@lysnikolaou @vadmium 

Just wondering, is there any update on this?

File: Doc/library/urllib.request.rst
Comment (eamanu) [reply Doc/library/urllib.request.rst]: I think that should be <protocol> because you are talking about *the* protocol, isn't?

**Review by eamanu — COMMENTED** [review]
I am not convinced about this changes. So, it will be better wait for another opinion :-)

**Comment by aixtools** [comment]
I do not know if a News entry is needed. If not, someone should add the "skip news" label.

As to the documentation modification, especially after the <type> - to make all of these clear I consider this a big improvement for "first-time" readers.

+1

**Comment by terryjreedy** [comment]
Add a blurb.

I completely agree that place holders should treated differently from literal text, but I don't know that <place-holder> is the proper form.

File: Doc/library/urllib.request.rst
Comment (merwok) [inline Doc/library/urllib.request.rst]: Missed one here

Comment (Denton-L) [reply Doc/library/urllib.request.rst]: Thanks!

**Comment by csabella** [comment]
I don't think there is a defined way to markup a placeholder.  Currently, `urllib.request` uses `<protocol>_` and `<scheme>_` in other parts of the document.  In addition, the argument clinic how-to uses a similar `<parameter_name>_length`.  Since a name can't start with `<`, this seems reasonable.

For `urllib.request` docs under the `BaseHandler Objects` section, the error method is defined as:
`BaseHandler.http_error_nnn(req, fp, code, msg, hdrs)` and then states `nnn should be a three-digit HTTP error code.`  

Interestingly, these methods listed under `OpenerDirector` are supposed to be defined as part of the  handler class passed in to OpenerDirector and those methods are defined (without the `<protocol>`) under `BaseHandler`.  IMHO, the listing of the methods in OpenerDirector should link to those definitions under BaseHandler.  

I also don't think the links are correct when BaseHandler or OpenerDirector are clicked anywhere within the doc.  They link to the class definition, which doesn't say much.  All the meat is under the `Objects` sections for each of those classes (which are also in the ToC on the left), so it would more helpful for links to go to those sections.  If that happened (linking the methods under OpenerDirector to BaseHandler), then BaseHandler could better explain that protocol is a placeholder.

**Comment by Denton-L** [comment]
Thanks for reviewing, @csabella!

>the error method is defined as:
`BaseHandler.http_error_nnn(req, fp, code, msg, hdrs)` and then states `nnn should be a three-digit HTTP error code.`

Thanks, I missed this. I'll change it to `http_error_<nnn>` in the next revision.

>IMHO, the listing of the methods in OpenerDirector should link to those definitions under BaseHandler.

Will do.

>All the meat is under the `Objects` sections for each of those classes (which are also in the ToC on the left), so it would more helpful for links to go to those sections.

Would it make sense to move the Objects stuff under their respective classes? Or would it make sense to link from the class to the specific Object section? I'm a little hesitant to change all of the links to point at the Object sections because I feel like it makes more logical sense to point to the actual class definition because, well, that's what they're actually referencing.

**Comment by csabella** [comment]
> Would it make sense to move the Objects stuff under their respective classes? Or would it make sense to link from the class to the specific Object section? I'm a little hesitant to change all of the links to point at the Object sections because I feel like it makes more logical sense to point to the actual class definition because, well, that's what they're actually referencing.

Hold off on this for now.  I only mentioned it because I found it frustrating that while reading the docs for urllib.request, I could never get to where I wanted to go!  :-)    While I think it is a worthwhile change, more discussion would most likely be required and I don't want it to delay your changes.

> IMHO, the listing of the methods in OpenerDirector should link to those definitions under BaseHandler.

However, I do think there is value in doing this part now as that can help clarify the issue raised by the bug report by explaining the <protocol> terminology under BaseHandler, where it belongs.

Thanks!

**Comment by Denton-L** [comment]
I'm currently trying to implement the links but it's surprisingly difficult. I'd appreciate a few pointers from someone who knows RST better than I do.

Here are a few of my attempts so far:

* Doesn't render a link

```rst
:meth:`BaseHandler.<protocol>_open`

.. method:: BaseHandler.<protocol>_open(req)
```

* Doesn't render a link

```rst
:meth:`BaseHandler.\<protocol\>_open`

.. method:: BaseHandler.\<protocol\>_open(req)
```

* We get ``ref:`BaseHandler.\<protocol\>_open <protocol_open>``` literally

```rst
``:ref:`BaseHandler.\<protocol\>_open <protocol_open>```

.. _protocol_open:

.. method:: BaseHandler.<protocol>_open(req)
```

* Also renders badly

```rst
:ref:```BaseHandler.\<protocol\>_open`` <protocol_open>`

.. _protocol_open:

.. method:: BaseHandler.<protocol>_open(req)
```

* Attempted to fix the above two issues with [this](https://stackoverflow.com/questions/45741961/sphinx-add-code-formatting-to-ref)

```rst
.. reference target not found

:any:`BaseHandler.<protocol>_open`

.. reference target not found

:any:`BaseHandler.\<protocol\>_open`

.. reference target not found

:any:`BaseHandler.\<protocol\>_open`

.. not formatted with inline code

:any:`BaseHandler.\<protocol\>_open <protocol_open>`

.. _protocol_open:

.. method:: BaseHandler.<protocol>_open(req)
```

**Comment by merwok** [comment]
I think you won’t be able to use function/method/reference roles with placeholders, there are constraints to the format of the contents inside the backticks.

I think the way here is to use a regular link and a substitution.


```
See this link to |Handler_protocol_handler|_

.. |Handler_protocol_handler|: replace:: ``Handler.{protocol}_handler``
.. _Handler_protocol_handler: target_ref_
```

(target_ref being the same text that is used in a ```:ref:`target_ref```` role to point to a section marked by `.. _target_ref:`)

http://docutils.sourceforge.net/docs/ref/rst/directives.html

**Comment by Denton-L** [comment]
@merwok Thanks for your help!

**Review by brettcannon — APPROVED** [review]

**Comment by brettcannon** [comment]
The wording and such SGTM, but I haven't verified the output looks correct (some fancy reST going on in this PR 😄 ). If someone can do that then I can merge this.

**Comment by Denton-L** [comment]
@brettcannon Thanks for reviewing. Here's a screenshot of that "fancy" part rendered in HTML:

image

**Comment by brettcannon** [comment]
Thanks @Denton-L !

**Pull request merged**
